core/src/main/java/org/pentaho/metaverse/graph/GraphCatalogWriter.java

```diff
@@ -0,0 +1,338 @@
+      }
+      // handles tables written to by a step
+      String resourceType = vertex.getProperty( DictionaryConst.PROPERTY_TYPE );
+      if ( propertyPopulated( resourceType ) && resourceType.equals( DictionaryConst.NODE_TYPE_DATA_TABLE ) ) {
+        String tableName = vertex.getProperty( DictionaryConst.PROPERTY_TABLE );
+        if ( propertyPopulated( tableName ) ) {
+          LineageDataResource dataResource = new LineageDataResource( tableName );
+          dataResource.setVertexId( vertex.getId() );
+          findDbConnectionProperties( vertex, dataResource, DictionaryConst.LINK_WRITESTO );
+          dataResource.setFields( getTableFields( tableName, graph ) );
+          dataResource.setDbSchema( vertex.getProperty( DictionaryConst.PROPERTY_SCHEMA ) );
+          outputTargets.add( dataResource );
+        }
+      }
+    } );
+
+    // Trace output fields to source fields
+    linkTargetFieldsToSources( outputTargets, inputSources, graph );
+
+    try {
+      lineageClient.processLineage( inputSources, outputTargets );
+    } catch ( Exception e ) {
+      log.error( e.getMessage(), e );
+    }
+
+    log.info( "Lineage processing done." );
+  }
+
+  private void findDbConnectionProperties( Vertex vertex, LineageDataResource dataResource, String readOrWrite ) {
+    Direction stepNodeDirection = readOrWrite.equals( DictionaryConst.LINK_READBY ) ? Direction.IN : Direction.OUT;
+    Direction stepEdgeDirection = readOrWrite.equals( DictionaryConst.LINK_READBY ) ? Direction.OUT : Direction.IN;
+    Iterator<Edge> queryEdges = vertex.getEdges( stepEdgeDirection ).iterator();
+    while ( queryEdges.hasNext() ) {
+      Edge queryEdge = queryEdges.next();
+      // find the step that reads this query
+      if ( queryEdge.getLabel().equals( readOrWrite ) &&
+        hasPropertyValue( queryEdge.getVertex( stepNodeDirection )
+          , DictionaryConst.PROPERTY_TYPE
+          , DictionaryConst.NODE_TYPE_TRANS_STEP ) ) {
+        Vertex tableStep = queryEdge.getVertex( stepNodeDirection );
+        Iterator<Edge> tableStepEdges = tableStep.getEdges( Direction.IN ).iterator();
+        while( tableStepEdges.hasNext() ) {
+          Edge tableEdge = tableStepEdges.next();
+          // find the DB connection that is a dependency of this table input step
+          if ( tableEdge.getLabel().equals( DictionaryConst.LINK_DEPENDENCYOF ) &&
+            hasPropertyValue( tableEdge.getVertex( Direction.OUT )
+              , DictionaryConst.PROPERTY_TYPE
+              , DictionaryConst.NODE_TYPE_DATASOURCE ) ) {
+            Vertex dbNode = tableEdge.getVertex( Direction.OUT );
+            dataResource.setDbHost( dbNode.getProperty( DictionaryConst.PROPERTY_HOST_NAME ) );
+            dataResource.setDbName( dbNode.getProperty( DictionaryConst.PROPERTY_DATABASE_NAME ) );
+          }
+        }
+      }
+    }
+  }
+
+  private String getSourceName( String fullName ) {
+    String sourceName = null;
+    if ( fullName.contains( "/" ) ) {
```

Comment:
Does the forward slash need to platform dependent?

Reply:
I don't know; I believe it depends on how Windows-stored files would be indexed in the catalog.